subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/topology_aware_scheduler.go

```diff
@@ -0,0 +1,298 @@
+			currentGpuIndices[searchGpuIndex] = availableGpuIndex
+			if searchGpuIndex == 0 {
+				currentAffinity[searchGpuIndex] = gpu
+			} else {
+				currentAffinity[searchGpuIndex] = findLCA(gpu, currentAffinity[searchGpuIndex-1])
+				// pruning: if the current LCA has been higher than the lowest ever,
+				// the node will be skipped
+				if (currentAffinity[searchGpuIndex] == nil && bestAffinity < highestLevel) ||
+					(currentAffinity[searchGpuIndex].GetLevel() > bestAffinity) {
+					availableGpuIndex++
+					continue
+				}
+			}
+			if searchGpuIndex == gpuNumber-1 {
+				foundOptimalAffinity := false
+				bestAffinity, foundOptimalAffinity = checkCurrentGpus(
+					currentAffinity[len(currentAffinity)-1].GetLevel(),
+					availableGpus,
+					currentGpuIndices,
+					bestAffinity,
+					bestAffinityGpus,
+					bestAffinityGpuIndices)
+				if foundOptimalAffinity {
+					// early stop: return if the solution is optimal (i.e., all buddies)
+					availableGpus = removePickedGpus(availableGpus, bestAffinityGpuIndices)
+					return bestAffinityGpus, availableGpus
+				}
+			} else {
+				searchGpuIndex++
+			}
+			availableGpuIndex++
+		}
+		searchGpuIndex--
+		if searchGpuIndex < 0 {
+			if bestAffinity == highestLevel {
+				panic(fmt.Sprintf("failed to allocate %v GPUs in picked node %v", gpuNumber, node.GetName()))
+			}
+			availableGpus = removePickedGpus(availableGpus, bestAffinityGpuIndices)
+			return bestAffinityGpus, availableGpus
+		}
+		availableGpuIndex = currentGpuIndices[searchGpuIndex] + 1
+	}
+}
+
+// checkCurrentGpus checks if the currently picked GPUs have the lowest LCA. It also checks if the solution
+// is optimal (if the GPUs are all buddies).
+func checkCurrentGpus(
+	affinity CellLevel,
+	gpus CellList,
+	currentIndices []int32,
+	bestAffinity CellLevel,
+	bestAffinityGpus CellList,
+	bestAffinityGpuIndices []int32) (CellLevel, bool) {
+
+	if affinity < bestAffinity {
+		copy(bestAffinityGpuIndices, currentIndices)
+		for i := 0; i < len(currentIndices); i++ {
+			bestAffinityGpus[i] = gpus[currentIndices[i]]
+		}
+		if affinity == bestAffinityGpus[0].GetLevel()+1 {
+			return affinity, true
+		} else {
+			return affinity, false
+		}
+	}
+	return bestAffinity, false
+}
+
+// removePickedGpus remove picked GPUs from the available GPU list.
+func removePickedGpus(gpus CellList, indices []int32) CellList {
+	for i, index := range indices {
+		offset := int32(i)
+		if i < len(indices)-1 {
+			nextIndex := indices[i+1]
+			copy(gpus[index-offset:nextIndex-offset-1], gpus[index+1:nextIndex])
+		} else {
+			copy(gpus[index-offset:], gpus[index+1:])
+		}
+	}
+	return gpus[:len(gpus)-len(indices)]
+}
+
+// findLCA finds the lowest common ancestor of two cells (nil if they have no LCA).
+func findLCA(lower Cell, higher Cell) Cell {
+	for lower.GetLevel() < higher.GetLevel() {
+		if lower.GetParent() == nil {
+			return nil
+		}
+		lower = lower.GetParent()
+	}
+	if lower == higher {
+		return lower
```

Comment:
Is it safe to compare the refer of cell?

Reply:
I think it is. We only create cell objects during initialization.
Another way is to create a unique identifier for each cell

---
In reply to: [335902545](https://github.com/microsoft/pai/pull/3684#discussion_r335902545) [](ancestors = 335902545)

Reply:
Changed to comparing cellName. 

---
In reply to: [335944279](https://github.com/microsoft/pai/pull/3684#discussion_r335944279) [](ancestors = 335944279,335902545)